# HW 2
## 학과: 경영학부
## 학번: 1985084
## 이름: 임수지
## 영상 링크: https://youtu.be/kRkHcgeVAsg?si=zU3BXCxJ6PQi5Zs-

### Problem 1

2차원 torch 배열(텐서)을 입력받아, 각 **행(row)** 의 합을 계산하여 리스트로 반환하는 함수를 작성하시오.  
반드시 **`for` 반복문**을 사용하여 합을 구해야 하며, `torch.sum()` 등의 내장 합산 함수를 사용해서는 안 된다.

#### 힌트
- `for row in A:` 구문으로 각 행을 순회할 수 있다.  
- `append()`를 이용해 리스트에 원소를 추가할 수 있다.  
- `element.item()`을 사용하면 tensor 형태의 값을 일반 숫자로 변환할 수 있다.


In [43]:
import torch

def row_sums_loop(A):
    """
    2차원 torch 텐서를 입력받아, 각 행의 합을 리스트로 반환하는 함수.
    반드시 for문을 사용하여 직접 계산해야 함.

    매개변수:
        A (torch.Tensor): (행, 열) 형태의 2차원 텐서

    반환값:
        list: 각 행의 합을 담은 리스트 (일반 숫자 형태)
    """
    row_sums = []  # 행의 합을 저장할 리스트
    
    # 각 행을 순회
    for row in A:
        total = 0  # 행의 합을 저장할 변수 초기화
        
        # 행의 각 원소를 순회하면서 합계 계산
        for element in row:
            total += element.item()  # tensor 값을 일반 숫자로 변환하여 더하기
        
        row_sums.append(total)  # 계산한 합을 리스트에 추가
    
    return row_sums


In [44]:
# ✅ 예시 실행
A = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

result = row_sums_loop(A)
print("입력 텐서:\n", A)
print("행 합계 결과:", result)


입력 텐서:
 tensor([[1, 2, 3],
        [4, 5, 6]])
행 합계 결과: [6, 15]


### Explanation

이 문제는 **반복문(for loop)** 을 이용해 2차원 텐서의 각 행을 순회하며  
그 안의 원소들을 하나씩 더해 나가는 기본적인 연습문제입니다.

1. `for row in A:`  
   → 2차원 텐서의 각 행을 하나씩 꺼냅니다.  
     예를 들어 `A`가  
     \[
     \begin{bmatrix}1 & 2 & 3\\ 4 & 5 & 6\end{bmatrix}
     \]
     이라면, 첫 번째 반복에서는 `[1, 2, 3]`, 두 번째 반복에서는 `[4, 5, 6]`이 `row`에 들어옵니다.

2. `for element in row:`  
   → 각 행 안의 원소를 하나씩 꺼내어 합을 계산합니다.  
     이때 `element`는 `tensor(3)`과 같은 **0차원 텐서(스칼라 텐서)** 형태이므로,  
     `.item()`을 이용해 일반 숫자로 바꾸어야 합니다.  

3. `total += element.item()`  
   → 변환된 숫자를 `total` 변수에 더합니다.

4. `row_sums.append(total)`  
   → 한 행의 합이 완성되면 이를 리스트에 추가합니다.



### Problem 2

2차원 torch 배열(텐서)을 입력받아, **중첩된 for문(nested loops)** 을 이용해  
행과 열을 바꾸어 놓은 **전치 행렬(transpose)** 을 직접 계산하는 함수를 작성하시오.  
(즉, 내장 함수 `A.T`나 `torch.transpose()`를 사용하지 말고,  
각 원소를 하나씩 옮겨서 새로운 텐서를 만들어야 한다.)



In [45]:

import torch

def transpose_loop(A):
    """
    2차원 torch 텐서 A의 전치행렬(transpose)을
    중첩된 for문을 이용해 직접 계산하는 함수.

    매개변수:
        A (torch.Tensor): (행, 열) 형태의 2차원 텐서
    
    반환값:
        torch.Tensor: A의 전치행렬 (열, 행) 형태
    """
    nrow, ncol = A.shape  # 원래 행과 열의 크기
    # 전치행렬은 (열, 행) 형태이므로 반대로 초기화
    A_T = torch.zeros((ncol, nrow), dtype=A.dtype)

    # 행, 열을 순회하며 위치를 바꾸어 채워 넣기
    for i in range(nrow):
        for j in range(ncol):
            A_T[j, i] = A[i, j]
    
    return A_T


In [46]:
# ✅ 예시 실행
A = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

result = transpose_loop(A)
print("입력 텐서:\n", A)
print("전치행렬 결과:\n", result)


입력 텐서:
 tensor([[1, 2, 3],
        [4, 5, 6]])
전치행렬 결과:
 tensor([[1, 4],
        [2, 5],
        [3, 6]])


### Explanation

이 문제는 **행(row)과 열(column)을 뒤바꾸는 전치 연산(transpose)** 을  
반복문으로 직접 구현하는 연습문제입니다.

1. `nrow, ncol = A.shape`  
   → 행렬의 행 개수와 열 개수를 구함.  
2. `A_T = torch.zeros((ncol, nrow))`  
   → 전치된 형태로 저장할 새로운 텐서를 0으로 초기화.  
     (전치행렬의 크기는 기존의 (열, 행) 크기임)  
3. 두 개의 for문을 이용하여 원소를 하나씩 옮김:
여기서 A[i, j]의 값은 전치행렬의 [j, i] 위치로 이동함.
4. 완성된 전치행렬 A_T를 반환함.


### Problem 3 https://youtu.be/kRkHcgeVAsg?si=zU3BXCxJ6PQi5Zs- 영상 0:00~1:43

1차원 torch 텐서를 입력받아 **Run-Length Encoding (RLE)** 을 수행하는 함수를 작성하시오.  
RLE는 연속해서 같은 값이 반복될 때,  
그 값을 한 번만 기록하고 반복 횟수를 함께 저장하는 압축 방식이다.

함수는 두 개의 Python **리스트**를 반환해야 한다:
- `values`: 연속 구간의 고유 값들
- `counts`: 해당 값이 연속된 횟수들  

반환되는 리스트에는 torch 텐서가 아닌 **일반 파이썬 숫자**만 포함되어야 한다.  
(`tensor(3)`을 `3`으로 바꾸려면 `.item()`을 사용)



In [47]:

import torch

def run_length_encode(seq):
    """
    1차원 torch 텐서에 대해 Run-Length Encoding(RLE)을 수행하는 함수.

    매개변수:
        seq (torch.Tensor): 1차원 텐서 (예: tensor([1,1,2,2,3]))

    반환값:
        values (list): 연속 구간의 값들을 담은 리스트 (파이썬 숫자)
        counts (list): 각 값의 반복 횟수를 담은 리스트
    """
    values = []  # 고유 값 저장
    counts = []  # 각 값의 반복 횟수 저장

    if len(seq) == 0:
        return values, counts  # 빈 텐서일 경우 빈 리스트 반환

    # 첫 번째 원소를 기준으로 시작
    current_value = seq[0].item()
    count = 1

    # 두 번째 원소부터 끝까지 순회
    for i in range(1, len(seq)):
        val = seq[i].item()
        if val == current_value:
            count += 1  # 같은 값이면 카운트 증가
        else:
            # 값이 바뀌면 저장 후 초기화
            values.append(current_value)
            counts.append(count)
            current_value = val
            count = 1

    # 마지막 구간도 저장
    values.append(current_value)
    counts.append(count)

    return values, counts


In [48]:
# ✅ 예시 실행
seq = torch.tensor([1,1,1,2,2,3,3,3,3,2])
values, counts = run_length_encode(seq)

print("입력 시퀀스:", seq.tolist())
print("values:", values)
print("counts:", counts)


입력 시퀀스: [1, 1, 1, 2, 2, 3, 3, 3, 3, 2]
values: [1, 2, 3, 2]
counts: [3, 2, 4, 1]


### Explanation

이 문제는 **반복문과 조건문을 이용한 기본 알고리즘 구현 능력**을 평가하는 문제입니다.  
`torch`의 고급 함수 대신 순수한 논리 흐름으로 문제를 해결해야 합니다.

1. 초기화
→ 첫 번째 원소를 현재 구간의 값으로 설정하고 카운트를 1로 시작합니다.

2. 반복문으로 순회
→ 이전 값과 같으면 카운트를 늘리고,
다르면 기존 구간을 리스트에 저장하고 새 구간을 시작합니다.

3. 마지막 구간 저장
4. 결과 반환


### Problem 4 https://youtu.be/kRkHcgeVAsg?si=zU3BXCxJ6PQi5Zs- 영상 1:48~3:31

PyTorch에서 `backward()`를 같은 변수에 여러 번 호출하면,  
이전의 gradient 값이 **누적(accumulate)** 되어 `x.grad`에 더해집니다.  
즉, 수학적으로는 새로운 도함수를 구하는 것처럼 보이지만,  
PyTorch에서는 기본적으로 **기존의 gradient에 더하는 방식**으로 동작합니다.

따라서 새로운 gradient를 처음부터 다시 계산하고 싶다면,  
반드시 `x.grad.zero_()`로 기존 값을 **초기화(reset)** 해야 합니다.


In [49]:
import torch

# (i) x 정의 및 z = x**2 계산
x = torch.tensor([2.0], requires_grad=True)
z = x**2

# (ii) 첫 번째 backward 호출
z.backward()
print("첫 번째 backward 후:", x.grad)

# (iii) 기존 grad를 초기화하지 않고 다시 backward 호출
z = x**2
z.backward()
print("두 번째 backward 후:", x.grad)

# (iv) grad 초기화 후 다시 계산
x.grad.zero_()  # 기존 gradient 제거
z = x**2
z.backward()
print("zero_() 후 backward:", x.grad)


첫 번째 backward 후: tensor([4.])
두 번째 backward 후: tensor([8.])
zero_() 후 backward: tensor([4.])


### Explanation

| 단계 | 코드 요약 | 기대 출력 | 설명 |
|:--|:--|:--|:--|
| (i) | `x = torch.tensor([2.0], requires_grad=True)` <br> `z = x**2` | - | 학습 가능한 변수 x 정의 |
| (ii) | `z.backward()` | `x.grad = tensor([4.])` | \(\frac{dz}{dx} = 2x = 4\) |
| (iii) | `z = x**2; z.backward()` (초기화 안 함) | `x.grad = tensor([8.])` | 이전 4에 새 4가 더해져 총 8 |
| (iv) | `x.grad.zero_()` 후 다시 `z.backward()` | `x.grad = tensor([4.])` | 초기화 후 새 gradient만 반영됨 |

---

#### 2️⃣ 왜 두 번째 backward 결과가 더 큰가?

PyTorch에서는 `backward()` 호출 시 기존 gradient를 지우지 않고 **누적(add)** 하기 때문입니다.  
그래서 첫 번째 backward에서 얻은 값 4가 그대로 남아 있고,  
두 번째 backward에서 새로 계산된 4가 더해져 최종적으로 8이 됩니다.

---

#### 3️⃣ `zero_()`를 이용한 초기화

`x.grad.zero_()`는 `x.grad` 텐서의 값을 0으로 바꿉니다.  
이후 `backward()`를 호출하면 새로운 gradient만 저장됩니다.  
이 방식은 “기존의 gradient는 버리고 새로 시작”할 때 필요합니다.

---

#### 4️⃣ (v) PyTorch가 기본적으로 gradient를 누적하는 이유

PyTorch는 **mini-batch 학습(batch training)** 을 지원하기 위해  
gradient를 기본적으로 **누적(accumulate)** 하는 구조를 가집니다.

- 신경망 학습에서는 하나의 batch(예: 데이터 32개)에 대해  
  forward → backward를 수행한 후,  
  다음 batch의 gradient를 **이전 gradient에 더하는 방식**으로  
  전체 데이터의 평균적인 gradient를 계산합니다.
- 즉, `x.grad`는 여러 batch의 gradient 합을 유지하게 되므로,  
  **전체 데이터셋의 gradient 근사치**를 쉽게 구할 수 있습니다.

따라서 PyTorch는 gradient를 덮어쓰지 않고 누적하는 것이 기본 동작이며,  
이로 인해 개발자는 여러 batch의 기울기를 직접 합산할 필요가 없습니다.



### Problem 5 https://youtu.be/kRkHcgeVAsg?si=zU3BXCxJ6PQi5Zs- 영상 3:32~6:05

다음과 같이 정의하자.
\[
x = 2.0,\quad
y = x^3,\quad
z = 2y + \sin(y)
\]

PyTorch에서 `requires_grad=True`로 텐서를 생성하면,  
그 이후 모든 계산(`x**3`, `sin(y)` 등)을 **계산 그래프(computation graph)** 로 기록합니다.  
이 그래프를 이용해 `.backward()`를 호출하면,  
계산 과정을 역추적하며 자동으로 미분을 수행할 수 있습니다.

그러나 이 계산 그래프는 메모리를 많이 차지하므로,  
PyTorch는 기본적으로 한 번 `backward()`를 호출하면 그래프를 **자동으로 삭제(free)** 합니다.  
만약 동일한 그래프를 다시 사용하고 싶다면,  
`retain_graph=True` 옵션을 주어 그래프를 **유지(keep)** 해야 합니다.


In [50]:
import torch

# (i) x, y, z 정의
x = torch.tensor([2.0], requires_grad=True)
y = x**3
z = 2*y + torch.sin(y)
print("x =", x.item(), "  y =", y.item(), "  z =", z.item())

# (ii) z.backward() 호출 후 x.grad 출력 → dz/dx 계산
z.backward()
print("dz/dx at x=2 :", x.grad)

# (iii) y.backward() 호출 (z.backward 이후)
try:
    y.backward()
except Exception as e:
    print("Error message:", e)


x = 2.0   y = 8.0   z = 16.98935890197754
dz/dx at x=2 : tensor([22.2540])
Error message: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.


### 💬 (iii) 해설

이때 오류가 발생합니다.

> RuntimeError: Trying to backward through the graph a second time...

왜냐하면 `z.backward()` 이후 그래프가 이미 메모리에서 해제되었기 때문입니다.  
따라서 같은 계산 그래프를 사용해 다시 `backward()`를 호출할 수 없습니다.

In [51]:
# (iv) retain_graph=True 사용 예시
import torch

# 계산 그래프 다시 생성
x = torch.tensor([2.0], requires_grad=True)
y = x**3
z = 2*y + torch.sin(y)

# (A) 그래프를 유지하면서 z.backward() 실행
z.backward(retain_graph=True)
print("dz/dx at x=2 :", x.grad)

# (B) 기존 gradient 제거
x.grad.zero_()

# (C) 같은 그래프에서 y.backward() 실행 → dy/dx 계산
y.backward()
print("dy/dx at x=2 :", x.grad)


dz/dx at x=2 : tensor([22.2540])
dy/dx at x=2 : tensor([12.])


### 💬 (iv) 해설

| 코드 구간 | 설명 |
|------------|------|
| (A) `retain_graph=True` | 첫 번째 `backward()` 시 그래프가 해제되지 않도록 유지함. <br>이후 또 다른 `backward()`를 실행할 수 있음. |
| (B) `x.grad.zero_()` | 기존의 gradient(이전 `dz/dx` 값)를 0으로 초기화. <br>이 과정을 생략하면 `dy/dx`가 기존 gradient 위에 누적되어 잘못된 값이 나옴. |
| (C) `y.backward()` | 같은 그래프를 이용해 `dy/dx` 계산. \(dy/dx = 3x^2 = 12\). |

만약 (B)의 `x.grad.zero_()`를 생략하면,  
기존의 `dz/dx`가 남아 있어 `dy/dx`가 **누적(accumulate)** 되어 더 큰 값이 나옵니다.  
따라서 항상 새로운 미분을 계산하기 전에는 `.zero_()`로 초기화해야 합니다.


In [52]:
# (v) 그래프를 새로 만드는 방법 (rebuild 방식)

# Pass 1: dz/dx 계산
import torch
x = torch.tensor([2.0], requires_grad=True)
y = x**3
z = 2*y + torch.sin(y)
z.backward()
print("dz/dx (fresh graph):", x.grad)

# Pass 2: 그래프 새로 만들어 dy/dx 계산
x = torch.tensor([2.0], requires_grad=True)
y = x**3
y.backward()
print("dy/dx (fresh graph):", x.grad)


dz/dx (fresh graph): tensor([22.2540])
dy/dx (fresh graph): tensor([12.])


### 💬 (v) 해설

이 방식은 `retain_graph=True`를 쓰지 않고,  
**필요할 때마다 계산 그래프를 새로 만드는 방식**입니다.

| 코드 구간 | 설명 |
|------------|------|
| Pass 1 | `dz/dx` 계산을 위한 그래프를 생성 |
| Pass 2 | 새로운 `x`와 `y`로 그래프를 다시 만들고 `dy/dx` 계산 |

이 방식의 장점은 불필요한 그래프 유지로 인한 메모리 낭비가 없다는 점입니다.  
다만, 같은 계산을 여러 번 반복해야 하므로 연산 비용이 늘어날 수 있습니다.


In [53]:
# (vi) zero_()를 이용한 대안적 방식

# Pass 1: dz/dx 계산
import torch
x = torch.tensor([2.0], requires_grad=True)
y = x**3
z = 2*y + torch.sin(y)
z.backward()
print("dz/dx (fresh graph):", x.grad)

# Pass 2: zero_()로 초기화 후 새 계산
x.grad.zero_()
y = x**3
y.backward()
print("dy/dx (fresh graph):", x.grad)


dz/dx (fresh graph): tensor([22.2540])
dy/dx (fresh graph): tensor([12.])


### 💬 (vi) 해설

| 코드 구간 | 설명 |
|------------|------|
| Pass 1 | `dz/dx` 계산 (새 그래프). |
| `x.grad.zero_()` | 기존 gradient 제거. <br>새로운 미분값만 반영되도록 초기화. |
| Pass 2 | 새 `y = x**3` 계산 후 `y.backward()` → `dy/dx = 12` |

이 방식은 **그래프를 다시 생성하지 않고** 동일한 변수 `x`를 재사용할 수 있다는 장점이 있습니다.  
하지만 `y = x**3` 등의 계산을 다시 실행해야 하므로,  
결국 그래프를 내부적으로 다시 구성합니다.


### Problem 6


In [54]:
import torch
from torch.utils.data import TensorDataset, DataLoader

# 데이터 및 파라미터 설정
X_vals = torch.tensor([1,3,2,5,0, 2,1,3,5,0], dtype=torch.float32)
theta  = torch.tensor([1.0], dtype=torch.float32, requires_grad=True)

# Dataset / DataLoader (mini-batch 크기 = 5)
ds  = TensorDataset(X_vals)
ldr = DataLoader(ds, batch_size=5, shuffle=False)

total_ll = 0.0  # 전체 로그우도 합

for (batch,) in ldr:
    # 배치 로그우도 계산
    ll_batch = torch.sum(batch * torch.log(theta) - theta - torch.lgamma(batch + 1.0))
    total_ll += ll_batch.item()

    # backward() 호출 → gradient 누적
    ll_batch.backward()
    print("log-likelihood (누적):", total_ll)
    print("누적된 gradient theta.grad:", theta.grad)


log-likelihood (누적): -12.272397994995117
누적된 gradient theta.grad: tensor([6.])
log-likelihood (누적): -24.544795989990234
누적된 gradient theta.grad: tensor([12.])


### (i) 각 반복(iteration)에서의 `ll_batch` 및 `total_ll` 값

**첫 번째 batch (X₁~X₅ = [1,3,2,5,0])**
\[
\ell_1(\theta=1) = \sum (x_i \log 1 - 1 - \log(x_i!))
= \sum (-1 - \log(x_i!))
\]

`log(1)=0`이므로, 단순히 `-1 - log(x_i!)`의 합입니다.  
대략적인 값은 다음과 같습니다.
\[
\ell_1(1) \approx -1 - (0 + 1.099 + 0.693 + 4.787 + 0) = -7.58
\]

**두 번째 batch (X₆~X₁₀ = [2,1,3,5,0])**
\[
\ell_2(1) \approx -1 - (0.693 + 0 + 1.099 + 4.787 + 0) = -7.58
\]

즉, `ll_batch` 값은 두 배치 모두 약 `-7.58` 정도이며,  
누적합 `total_ll`은 다음과 같이 갱신됩니다.

| 반복 | 배치 데이터 | ll_batch | total_ll |
|------|--------------|-----------|-----------|
| 1회차 | [1,3,2,5,0] | ≈ -7.58 | ≈ -7.58 |
| 2회차 | [2,1,3,5,0] | ≈ -7.58 | ≈ -15.16 |



### (ii) 각 반복에서의 `theta.grad` 값

점수함수(score function)는  
\[
\frac{\partial \ell(\theta)}{\partial \theta}
= \sum_{i=1}^{n} \left(\frac{X_i}{\theta} - 1\right)
\]

따라서 \(\theta=1\)일 때,

- **첫 번째 배치**
  \[
  \sum_{i=1}^{5} (X_i - 1)
  = (1-1) + (3-1) + (2-1) + (5-1) + (0-1) = 6
  \]
  → `theta.grad = tensor([6.])`

- **두 번째 배치**
  \[
  \sum_{i=6}^{10} (X_i - 1)
  = (2-1) + (1-1) + (3-1) + (5-1) + (0-1) = 6
  \]
  PyTorch는 `.backward()`를 호출할 때마다 **gradient를 누적(accumulate)** 하므로  
  두 번째 backward 이후에는
  \[
  \text{누적된 } \theta.grad = 6 + 6 = 12
  \]

| 반복 | 배치 데이터 | θ.grad(현재 배치) | θ.grad(누적 결과) |
|------|--------------|--------------------|-------------------|
| 1회차 | [1,3,2,5,0] | 6 | 6 |
| 2회차 | [2,1,3,5,0] | 6 | 12 |


### (iii) 전체 데이터에 대한 점수함수 (총 gradient)

전체 데이터 10개에 대해,
\[
\left.\frac{\partial \ell(\theta)}{\partial \theta}\right|_{\theta=1}
= \sum_{i=1}^{10} (X_i - 1)
\]
\[
= (1-1)+(3-1)+(2-1)+(5-1)+(0-1)+(2-1)+(1-1)+(3-1)+(5-1)+(0-1)
= 12
\]

따라서 코드에서 마지막 출력되는 `theta.grad` 값은
\[
\boxed{\theta.grad = tensor([12.])}
\]

이는 PyTorch가 **각 배치별 gradient를 자동으로 더해주는 누적 기능** 덕분에,  
전체 데이터의 점수함수를 한 번에 계산할 수 있음을 보여줍니다.


### Problem 7

In [55]:
import torch

# (1) 변수 정의
x1 = torch.tensor([1.0], requires_grad=True)
x2 = torch.tensor([0.5], requires_grad=True)

# (2) f1, f2 계산
f1 = x1**2 + torch.sin(x2)
f2 = x1 * x2 + torch.exp(x1)

# (3) 두 출력을 하나의 벡터로 쌓기
F = torch.stack([f1, f2])

# (4) 첫 번째 출력 f1에 대해 backward() 수행 → J의 첫 번째 행 계산
F[0].backward(retain_graph=True)  # 그래프 유지 (두 번째 backward를 위해)
row1 = torch.tensor([x1.grad.item(), x2.grad.item()])

print("첫 번째 행 (∂f1/∂x1, ∂f1/∂x2):", row1)

# (5) gradient 초기화
x1.grad.zero_()
x2.grad.zero_()

# (6) 두 번째 출력 f2에 대해 backward() 수행 → J의 두 번째 행 계산
F[1].backward()
row2 = torch.tensor([x1.grad.item(), x2.grad.item()])

print("두 번째 행 (∂f2/∂x1, ∂f2/∂x2):", row2)

# (7) 두 행을 합쳐서 최종 Jacobian 구성
J = torch.stack([row1, row2])
print("\n최종 Jacobian 행렬:\n", J)


첫 번째 행 (∂f1/∂x1, ∂f1/∂x2): tensor([2.0000, 0.8776])
두 번째 행 (∂f2/∂x1, ∂f2/∂x2): tensor([3.2183, 1.0000])

최종 Jacobian 행렬:
 tensor([[2.0000, 0.8776],
        [3.2183, 1.0000]])


### Explanation

| 단계 | 코드 | 설명 |
|------|------|------|
| (1) | `x1, x2`를 `requires_grad=True`로 정의 | 자동미분 가능한 변수 생성 |
| (2) | `f1 = x1**2 + sin(x2)` <br> `f2 = x1*x2 + exp(x1)` | 함수 정의 |
| (3) | `F = torch.stack([f1, f2])` | 두 출력을 하나의 텐서로 묶음 |
| (4) | `F[0].backward(retain_graph=True)` | 첫 번째 출력에 대해 미분 수행 |
| (5) | `x1.grad.zero_(), x2.grad.zero_()` | 기존 gradient 초기화 |
| (6) | `F[1].backward()` | 두 번째 출력에 대해 미분 수행 |
| (7) | `torch.stack([row1, row2])` | 두 행을 쌓아 최종 Jacobian 생성 |

---

#### 2️⃣ 이론적 계산 확인

\[
\begin{align*}
f_1(x_1, x_2) &= x_1^2 + \sin(x_2) \\
f_2(x_1, x_2) &= x_1 x_2 + e^{x_1}
\end{align*}
\]

따라서,
\[
J_F(x_1, x_2) =
\begin{bmatrix}
2x_1 & \cos(x_2) \\
x_2 + e^{x_1} & x_1
\end{bmatrix}
\]

\[
x_1 = 1.0, \quad x_2 = 0.5
\]
를 대입하면,

\[
J_F(1, 0.5) =
\begin{bmatrix}
2(1) & \cos(0.5) \\
0.5 + e^{1} & 1
\end{bmatrix}
=
\begin{bmatrix}
2.0 & 0.8776 \\
3.2183 & 1.0
\end{bmatrix}
\]

---

#### 3️⃣ PyTorch 계산 결과 확인

실행 결과 예시는 다음과 같습니다.

첫 번째 행 (∂f1/∂x1, ∂f1/∂x2): tensor([2.0000, 0.8776])
두 번째 행 (∂f2/∂x1, ∂f2/∂x2): tensor([3.2183, 1.0000])

최종 Jacobian 행렬:
tensor([[2.0000, 0.8776],
[3.2183, 1.0000]])


이는 이론적 결과와 완전히 일치합니다.


### Problem 8 https://youtu.be/kRkHcgeVAsg?si=zU3BXCxJ6PQi5Zs- 영상 6:06~8:53

정재를 포함한 11명의 참가자가 노래에 맞춰 무작위로 돌다가  
노래가 멈추면 **2명씩 짝을 지어**, **한 명이 남으면 탈락**하는 게임이다.  

이 확률을 컴퓨터 시뮬레이션으로 확인하기 위해 아래 코드를 완성하시오.


In [56]:
import torch
from torch.utils.data import TensorDataset, DataLoader

# 11명의 참가자 (0~10번), 정재는 0번
N = 11
num_trials = 10000
participants = torch.arange(N)
ds = TensorDataset(participants)
success = 0

for t in range(num_trials):
    # 🔹 (정답) DataLoader 생성: 참가자 순서를 랜덤으로 섞고(batch_size=2)
    loader = DataLoader(ds, batch_size=2, shuffle=True, drop_last=False)
    survived = True

    for (batch,) in loader:
        # 한 명만 남으면 탈락
        if batch.shape[0] == 1:
            # 마지막 남은 사람이 정재(0)면 탈락 처리
            if batch.item() == 0:
                survived = False

    if survived:
        success += 1

print("시뮬레이션 생존 확률:", success / num_trials)


시뮬레이션 생존 확률: 0.9086


### 정답:
DataLoader(ds, batch_size=2, shuffle=True, drop_last=False)

### Explanation

1. **`participants = torch.arange(N)`**
   → 참가자 번호 `[0,1,2,...,10]`을 생성합니다.  
   여기서 **0번이 정재**입니다.

2. **`TensorDataset(participants)`**
   → 참가자 번호를 하나의 텐서 데이터셋으로 만듭니다.  
   (DataLoader는 Dataset 형태를 입력으로 받음)

3. **`DataLoader(ds, batch_size=2, shuffle=True)`**
   → DataLoader는 매 trial마다 11명을 **랜덤하게 섞고**,  
     2명씩 짝을 짓습니다.  
     단, `drop_last=False`이므로 11명이 홀수일 때 마지막 한 명은 그대로 남습니다.

4. **내부 for문 (`for (batch,) in loader:`)**
   - 각 `batch`는 두 명의 참가자 (예: `[7, 4]`, `[3, 1]`, ...)  
   - 만약 `batch.shape[0] == 1`이면, 마지막에 **혼자 남은 사람**입니다.  
   - 그 사람이 `0`이면 → 정재 탈락(`survived=False`).

5. **`if survived:`**
   → 정재가 탈락하지 않았으면 생존 횟수(`success`)를 1 증가시킵니다.

6. **`success / num_trials`**
   → 전체 시뮬레이션 중 생존한 비율을 계산합니다.

### 요약

| 항목 | 내용 |
|------|------|
| 실험 목적 | PyTorch DataLoader의 `shuffle`과 `batch_size`를 이용해 랜덤 짝짓기 시뮬레이션 |
| 핵심 아이디어 | 정재가 마지막 한 명으로 남는 경우만 탈락 |
| 이론적 확률 | \(10/11 ≈ 0.9091\) |
| 시뮬레이션 결과 | 약 0.909 근처 |
| 개념 포인트 | `DataLoader`의 무작위 샘플링과 gradient 누적이 아닌 반복 샘플링 개념 이해 |